src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,494 @@
+  const { activeSlideIndex: userActiveIndex = 0, children, ...rest } = props;
+
+  // Generate a stateful random id if not specified
+  const [id] = React.useState(
+    () => props.id ?? `iui-carousel-${getRandomValue(10)}`,
+  );
+
+  useTheme();
+
+  const [currentIndex, setCurrentIndex] = React.useState(userActiveIndex);
+  React.useEffect(() => {
+    setCurrentIndex(userActiveIndex);
+  }, [userActiveIndex]);
+
+  const [slideCount, setSlideCount] = React.useState(0);
+
+  const [keysPressed, setKeysPressed] = React.useState<Record<string, boolean>>(
+    {},
+  );
+
+  const isManuallyUpdating = React.useRef(false);
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    if (event.altKey || event.ctrlKey || event.metaKey || event.shiftKey) {
+      return;
+    }
+
+    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      setKeysPressed((old) => ({ ...old, [event.key]: true }));
+    }
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent) => {
+    switch (event.key) {
+      case 'ArrowLeft': {
+        isManuallyUpdating.current = true;
+        setKeysPressed((old) => ({ ...old, ArrowLeft: false }));
+        setCurrentIndex((old) => (slideCount + old - 1) % slideCount);
+        break;
+      }
+      case 'ArrowRight': {
+        isManuallyUpdating.current = true;
+        setKeysPressed((old) => ({ ...old, ArrowRight: false }));
+        setCurrentIndex((old) => (slideCount + old + 1) % slideCount);
+        break;
+      }
+    }
+  };
+
+  return (
+    <section
+      className='iui-carousel'
+      aria-roledescription='carousel'
+      tabIndex={0}
+      onKeyDown={handleKeyDown}
+      onKeyUp={handleKeyUp}
+      {...rest}
+      id={id}
+    >
+      <CarouselContext.Provider
+        value={{
+          currentIndex,
+          setCurrentIndex,
+          slideCount,
+          setSlideCount,
+          keysPressed,
+          idPrefix: id,
+          isManuallyUpdating,
+        }}
+      >
+        {children}
+      </CarouselContext.Provider>
+    </section>
+  );
+};
+
+/**
+ * CarouselSlider is the scrollable list that should consist of CarouselSlide components.
+ */
+const CarouselSlider = React.forwardRef<
+  HTMLOListElement,
+  React.ComponentPropsWithoutRef<'ol'>
+>((props, ref) => {
+  const context = React.useContext(CarouselContext);
+
+  if (!context) {
+    throw new Error('CarouselSlider must be used within Carousel');
+  }
+
+  const { currentIndex, setSlideCount, idPrefix } = context;
+  const sliderRef = React.useRef<HTMLOListElement>(null);
+  const refs = useMergedRefs(sliderRef, ref);
+
+  React.useLayoutEffect(() => {
+    setSlideCount(React.Children.count(props.children));
+  }, [props.children, setSlideCount]);
+
+  const items = React.useMemo(
+    () =>
+      React.Children.map(props.children, (child, index) =>
+        React.isValidElement(child)
+          ? React.cloneElement(child, {
+              id: `${idPrefix}--slide-${index}`,
```

Comment:
So if user sets `id` on `CarouselSlide`, it will be overwritten. Because `CarouselSlide` props extends default `li` props.

Reply:
So if we take `id` from user, then we need to somehow set in the `aria-controls` of dots, which adds complexity. Do you think it's worth it?